File: jre/src/main/java/com/azure/data/cosmos/serialization/hybridrow/SchemaId.java
//------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
//------------------------------------------------------------

package com.azure.data.cosmos.serialization.hybridrow;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
import com.fasterxml.jackson.databind.exc.MismatchedInputException;
import com.fasterxml.jackson.databind.ser.std.StdSerializer;

import java.io.IOException;

import static com.google.common.base.Strings.lenientFormat;

/**
 * The unique identifier for a schema
 * Identifiers must be unique within the scope of the database in which they are used.
 */
@JsonDeserialize(using = SchemaId.JsonDeserializer.class)
@JsonSerialize(using = SchemaId.JsonSerializer.class)
public final class SchemaId {

    // TODO: DANOBLE: Consider caching SchemaId instances to reduce memory footprint

    public static final SchemaId INVALID = null;
    public static final SchemaId NONE = new SchemaId(-1);
    public static final int SIZE = (Integer.SIZE / Byte.SIZE);

    private static long MAX_VALUE = 0x00000000FFFFFFFFL;
    private final int id;

    /**
     * Initializes a new instance of the {@link SchemaId} struct.
     *
     * @param id The underlying globally unique identifier of the schema.
     */
    private SchemaId(int id) {
        this.id = id;
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof SchemaId && this.equals((SchemaId) other);
    }

    /**
     * {@code true} if this is the same {@link SchemaId} as {@code other}
     *
     * @param other The value to compare against.
     * @return True if the two values are the same.
     */
    public boolean equals(SchemaId other) {
        if (null == other) {
            return false;
        }
        return this.id() == other.id();
    }

    @Override
    public int hashCode() {
        return Integer.valueOf(this.id()).hashCode();
    }

    /**
     * The underlying integer value of this {@link SchemaId}
     *
     * @return The integer value of this {@link SchemaId}
     */
    public int id() {
        return this.id;
    }

    /**
     * Returns a {@link SchemaId} from a specified underlying integer value
     *
     * @return The integer value of this {@link SchemaId}
     */
    public static SchemaId from(int id) {
        return new SchemaId(id);
    }

    @Override
    public String toString() {
        return String.valueOf(this.id());
    }

    static final class JsonDeserializer extends StdDeserializer<SchemaId> {

        private JsonDeserializer() {
            super(SchemaId.class);
        }

        @Override
        public SchemaId deserialize(final JsonParser parser, final DeserializationContext context) throws IOException, JsonProcessingException {

            final long value = parser.getLongValue();

            if (value < 0 || value > MAX_VALUE) {
                String message = lenientFormat("expected integer value in [0, 4294967295], not %s", value);
                throw MismatchedInputException.from(parser, SchemaId.class, message);
            }

            return new SchemaId((int) value);
        }
    }

    static final class JsonSerializer extends StdSerializer<SchemaId> {

        private JsonSerializer() {
            super(SchemaId.class);
        }

        @Override
        public void serialize(final SchemaId value, final JsonGenerator generator, final SerializerProvider provider) throws IOException {
            generator.writeNumber((long) value.id() & MAX_VALUE);
        }
    }
}

File: src/Serialization/HybridRow/SchemaId.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

#pragma warning disable IDE0041 // Use 'is null' check
#pragma warning disable IDE0070 // Use 'System.HashCode'

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow
{
    using System;
    using System.Diagnostics;
    using System.Diagnostics.CodeAnalysis;
    using System.Runtime.InteropServices;
    using Microsoft.Azure.Cosmos.Core;
    using Newtonsoft.Json;

    /// <summary>The unique identifier for a schema.</summary>
    /// <remarks>Identifiers must be unique within the scope of the database in which they are used.</remarks>
    [JsonConverter(typeof(SchemaIdConverter))]
    [DebuggerDisplay("{" + nameof(SchemaId.Id) + "}")]
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public readonly struct SchemaId : IEquatable<SchemaId>
    {
        public const int Size = sizeof(int);
        public static readonly SchemaId Invalid = default;

        /// <summary>Initializes a new instance of the <see cref="SchemaId" /> struct.</summary>
        /// <param name="id">The underlying globally unique identifier of the schema.</param>
        public SchemaId(int id)
        {
            this.Id = id;
        }

        /// <summary>The underlying identifier.</summary>
        public int Id { get; }

        /// <summary>Integer conversion operator.</summary>
        [SuppressMessage("Usage", "CA2225:Operator overloads have named alternates", Justification = "Constructor")]
        public static explicit operator SchemaId(int id)
        {
            return new SchemaId(id);
        }

        /// <summary>Integer conversion operator.</summary>
        [SuppressMessage("Usage", "CA2225:Operator overloads have named alternates", Justification = "Id property")]
        public static explicit operator int(SchemaId id)
        {
            return id.Id;
        }

        /// <summary>Operator == overload.</summary>
        public static bool operator ==(SchemaId left, SchemaId right)
        {
            return left.Equals(right);
        }

        /// <summary>Operator != overload.</summary>
        public static bool operator !=(SchemaId left, SchemaId right)
        {
            return !left.Equals(right);
        }

        /// <summary><see cref="object.Equals(object)" /> overload.</summary>
        public override bool Equals(object obj)
        {
            if (object.ReferenceEquals(null, obj))
            {
                return false;
            }

            return obj is SchemaId id && this.Equals(id);
        }

        /// <summary><see cref="object.GetHashCode" /> overload.</summary>
        public override int GetHashCode()
        {
            return this.Id.GetHashCode();
        }

        /// <summary>Returns true if this is the same <see cref="SchemaId" /> as <see cref="other" />.</summary>
        /// <param name="other">The value to compare against.</param>
        /// <returns>True if the two values are the same.</returns>
        public bool Equals(SchemaId other)
        {
            return this.Id == other.Id;
        }

        /// <summary><see cref="object.ToString" /> overload.</summary>
        public override string ToString()
        {
            return this.Id.ToString();
        }

        /// <summary>Helper class for parsing <see cref="SchemaId" /> from JSON.</summary>
        internal class SchemaIdConverter : JsonConverter
        {
            public override bool CanWrite => true;

            public override bool CanConvert(Type objectType)
            {
                return typeof(SchemaId).IsAssignableFrom(objectType);
            }

            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
            {
                Contract.Requires(reader.TokenType == JsonToken.Integer);
                return new SchemaId(checked((int)(long)reader.Value));
            }

            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
            {
                writer.WriteValue((long)((SchemaId)value).Id);
            }
        }
    }
}
